Rewrite index.js

asamuzaK opened the pull request:
Hello,

In [Toolkit version format](https://developer.mozilla.org/en-US/docs/Toolkit_version_format), version format is described as:

> A version string consists of one or more version parts, separated with dots.
> 
> Each version part is itself parsed as a sequence of four parts: <number-a><string-b><number-c><string-d>. Each of the parts is optional. Numbers are integers base 10 (may be negative), strings are non-numeric ASCII characters.

The expression now used in [mozilla-toolkit-versioning/lib/utils.js](https://github.com/jsantell/mozilla-toolkit-versioning/blob/master/lib/utils.js) is:
`/^([0-9\.]*)([a-zA-Z]*)([0-9\.]*)([a-zA-Z]*)$/`

But this expression does not match Mozilla's version part spec:
- Numbers may be negative.
- `non-numeric ASCII characters` are not only Alphabets.

2nd point.
[mozilla-toolkit-versioning/index.js](https://github.com/jsantell/mozilla-toolkit-versioning/blob/master/index.js) miss-handles some input cases.
The most typical case is that it can't handle `>1.2.*`.

I forked and rewritten index.js.
Please take a look at it.

Regards.

asamuzaK committed:
Rewrite index.js

asamuzaK committed:
removed needles `if else`

asamuzaK committed:
Moved expression out of loop

asamuzaK committed:
Avoid template strings

asamuzaK committed:
Revert increment (vString)

asamuzaK commented:
Ah...
I'll check the Travis Errors and come back again.
Closing for now.

asamuzaK committed:
fix increment (vString)

asamuzaK committed:
add VERSION_PART_CAPTURE constants

asamuzaK committed:
Revert "add VERSION_PART_CAPTURE constants"

This reverts commit 35f3658f0325f15f475bf21102094651b0ac515b.

asamuzaK committed:
Add VERSION_PART_CAPTURE constants

asamuzaK committed:
Add VERSION_PART_CAPTURE inside increment()

asamuzaK committed:
Add `\` to expression string

asamuzaK commented:
Reopened.

Test passed [Job #4.2 - asamuzaK/mozilla-toolkit-versioning - Travis CI](https://travis-ci.org/asamuzaK/mozilla-toolkit-versioning/jobs/107618384).
But still have Travis Error...

jsantell commented:
Just landed travis changes, some weird dependency stuff going on with installing mocha/chai on 0.8 that I'm not going to burn time on figuring out: https://github.com/jsantell/mozilla-toolkit-versioning/pull/3

Rerun the tests now, should have a clearer picture!

jsantell commented:
Haven't had a chance to look over the changes, but I'd imagine will need some tests for the new features (negative numbers, non-numeric ascii)

asamuzaK committed:
Add support for node 4, 5 in tests; remove 0.8

asamuzaK committed:
Add newline at end of file

asamuzaK commented:
I'm adding tests in [mozilla-toolkit-versioning/index.js at Update-test/index.js](https://github.com/asamuzaK/mozilla-toolkit-versioning/blob/Update-test/index.js/test/index.js)

Will you please take a look at it and check whether I'm doing things right?
And, if tests added are OK, should I merge test/index.js in this Pull Request?
Or create a new one for that?

Tests added are...
For parse(version) single value:
`testParse('1.2.-1', '1.2.-1', '1.2.-1');`
`testParse('>=1.2.-1', '1.2.-1', undefined);`
`testParse('<=1.2.-1', undefined, '1.2.-1');`
`testParse('>1.2.-1', '1.2.-1.1', undefined);`
`testParse('<1.2.-1', undefined, '1.2.-1.-1');`

`testParse('1.2.*', '1.2.*', '1.2.*');`
`testParse('>=1.2.*', '1.2.*', undefined);`
`testParse('<=1.2.*', undefined, '1.2.*');`
`testParse('>1.2.*', '1.2.*.1', undefined);`
`testParse('<1.2.*', undefined, '1.2.*.-1');`

`testParse('1..', '1..', '1..');`
`testParse('>=1..', '1..', undefined);`
`testParse('<=1..', undefined, '1..');`
`testParse('>1..', '1..1', undefined);`
`testParse('<1..', undefined, '1..-1');`

For increment(version):
`testInc('1.2.-1', '1.2.-1.1');`
`testInc('1.2.*', '1.2.*.1');`
`testInc('1..', '1..1');`

asamuzaK commented:
I'm adding tests in [mozilla-toolkit-versioning/index.js at Update-test/index.js](https://github.com/asamuzaK/mozilla-toolkit-versioning/blob/Update-test/index.js/test/index.js)

Will you please take a look at it and check whether I'm doing things right?
And, if tests added are OK, should I merge test/index.js in this Pull Request?
Or create a new one for that?

Tests added are...
For parse(version) single value:
`testParse('1.2.-1', '1.2.-1', '1.2.-1');`
`testParse('>=1.2.-1', '1.2.-1', undefined);`
`testParse('<=1.2.-1', undefined, '1.2.-1');`
`testParse('>1.2.-1', '1.2.-1.1', undefined);`
`testParse('<1.2.-1', undefined, '1.2.-1.-1');`

`testParse('1.2.*', '1.2.*', '1.2.*');`
`testParse('>=1.2.*', '1.2.*', undefined);`
`testParse('<=1.2.*', undefined, '1.2.*');`
`testParse('>1.2.*', '1.2.*.1', undefined);`
`testParse('<1.2.*', undefined, '1.2.*.-1');`

`testParse('1..', '1..', '1..');`
`testParse('>=1..', '1..', undefined);`
`testParse('<=1..', undefined, '1..');`
`testParse('>1..', '1..1', undefined);`
`testParse('<1..', undefined, '1..-1');`

For increment(version):
`testInc('1.2.-1', '1.2.-1.1');`
`testInc('1.2.*', '1.2.*.1');`
`testInc('1..', '1..1');`

asamuzaK commented:
I'm adding tests in [mozilla-toolkit-versioning/index.js at Update-test/index.js](https://github.com/asamuzaK/mozilla-toolkit-versioning/blob/Update-test/index.js/test/index.js).

Will you please take a look at it and check whether I'm doing things right?
And, if tests added are OK, should I merge test/index.js in this Pull Request?
Or create a new one for that?

Tests added are...
For parse(version) single value:
`testParse('1.2.-1', '1.2.-1', '1.2.-1');`
`testParse('>=1.2.-1', '1.2.-1', undefined);`
`testParse('<=1.2.-1', undefined, '1.2.-1');`
`testParse('>1.2.-1', '1.2.-1.1', undefined);`
`testParse('<1.2.-1', undefined, '1.2.-1.-1');`

`testParse('1.2.*', '1.2.*', '1.2.*');`
`testParse('>=1.2.*', '1.2.*', undefined);`
`testParse('<=1.2.*', undefined, '1.2.*');`
`testParse('>1.2.*', '1.2.*.1', undefined);`
`testParse('<1.2.*', undefined, '1.2.*.-1');`

`testParse('1..', '1..', '1..');`
`testParse('>=1..', '1..', undefined);`
`testParse('<=1..', undefined, '1..');`
`testParse('>1..', '1..1', undefined);`
`testParse('<1..', undefined, '1..-1');`

For increment(version):
`testInc('1.2.-1', '1.2.-1.1');`
`testInc('1.2.*', '1.2.*.1');`
`testInc('1..', '1..1');`

asamuzaK commented:
In [jpm xpi creates invalid install.rdf · Issue #461 · mozilla-jetpack/jpm](https://github.com/mozilla-jetpack/jpm/issues/461), there is a range expression like:

> `">=44.0 || <=49.*"`

Is this kind of expression  (using `||`) acceptable?
If so, I think I need to fix my patch.

asamuzaK committed:
Add require('mozilla-version-comparator') and some fixes

asamuzaK committed:
Revert "Add require('mozilla-version-comparator') and some fixes"

This reverts commit 4a50a3ccc7256294bafb2f2a231e6ddfa5d55010.

asamuzaK committed:
Add require('mozilla-version-comparator')

asamuzaK committed:
Revert "Add require('mozilla-version-comparator')"

This reverts commit 78c07cfd19494e71586d097da2ccce7fc1831eac.

asamuzaK committed:
Add require('mozilla-version-comparator');

asamuzaK committed:
Merge remote-tracking branch 'refs/remotes/jsantell/master'

asamuzaK committed:
Change for-loop to function parseMinMax()

asamuzaK committed:
Moved handling `*` case

asamuzaK commented:
I think the patch is almost ready.
Before, I want several points to be clarified (it's written in the patch as `NOTE`).

Handling the `*` case:
Q1: Maybe `return { min: undefined, max: '*' }` should be better?

Handling version range expression:
Q2: What is expected in `2.3.4 - 1.2.3` case? (L > R)
`if (L > R) then (min = R, max = L)`? (current behavior in this patch)

Q3: Is `>=1.2.3 - <=2.3.4` acceptable? (with COMPARATOR)
IMO it is better to accept that kind of inputs.
Then we can treat `-` as an optional range separator,
and syntax for the input expression will be simpler.
Input syntax will be:
`INPUT_SYNTAX = VERSION_STRING((\s+-)?\s+VERSION_STRING)?`
where:
`VERSION_STRING = (COMPARATOR)?VERSION_FORMAT`
`COMPARATOR = [><]=?`
`VERSION_FORMAT = (VERSION_PART)?(\.(VERSION_PART)?)*`
`VERSION_PART = ((-?[\d]+)?([!-\-\/:-~]+)?){2}`

Q4: Is `>=1.2.3 || <=2.3.4` acceptable? (with `||`, maybe `&&` too)
IMO it should be rejected.

Regards.

asamuzaK committed:
fix `input` argument detection

asamuzaK committed:
Add `exp` argument for parseMinMax()

asamuzaK committed:
update index.js test/index.js

asamuzaK committed:
fix code

asamuzaK committed:
Update test

asamuzaK committed:
Update test

asamuzaK committed:
Add test '1.2.3 - 1.2.*'

asamuzaK committed:
Add test ' 1.2.3 ' (containing spaces in input)

asamuzaK committed:
Add fallback for inconsistent inputs

asamuzaK committed:
fix fallback

asamuzaK committed:
fix fallback test

asamuzaK committed:
update fallback

asamuzaK committed:
remove fallback and throw in (min > max) case

asamuzaK committed:
fix typo

asamuzaK committed:
remove lib/

asamuzaK committed:
Merge pull request #1 from asamuzaK/dev

Dev

asamuzaK commented:
Pull Request is ready.
Please take a look at it.

`input` argument syntax:
`INPUT_SYNTAX = VERSION_STRING((\s+-)?\s+VERSION_STRING)?`

Where:
`VERSION_STRING = (COMPARATOR)?VERSION_FORMAT`
`COMPARATOR = [><]=?`
`VERSION_FORMAT = (VERSION_PART)?(\.(VERSION_PART)?)*`
`VERSION_PART = ((-?[\d]+)?([!-\-\/:-~]+)?){2}`

Exceptional `input` cases:
Case `'*'`:
Returns `{ min: undefined, max: '*' }`.
`*` should be treated as `infinity`.
But `min: infinity` does not make sense and may cause regression.
So `*` should only be set in max version.

Case `'-'`:
Returns `{ min: undefined, max: undefined }`.
Parsed as range `undefined - undefined`.

Case `''` (empty string):
Returns `{ min: undefined, max: undefined }`.
Parsed as single value `undefined`.

Case `'VERSION_FORMAT -'` (VERSION_STRING without COMPARATOR):
Returns `{ min:VERSION_FORMAT, max: undefined }`.
Parsed as range `VERSION_FORMAT - undefined`.
Also `'- VERSION_FORMAT'` returns `{ min: undefined, max: VERSION_FORMAT }`.

Some `input` cases which throws:
- When `input` argument is not string type.
- When `input` string does not match INPUT_SYNTAX.
- When range separator is not `-`.
- When range expression is inconsistent. e.g.) `1.2.3 >=2.3.4`

asamuzaK committed:
Fix conditional expression not to be an assignment

asamuzaK committed:
clean up code.

asamuzaK committed:
Revert "clean up code."

This reverts commit ce63b14a462742d4b9279d42651d99b3dc9c6f78.

asamuzaK committed:
remove spaces

asamuzaK committed:
fix increment('1.2.-1')

asamuzaK committed:
fix tes pattern

asamuzaK committed:
put `else if` together

asamuzaK committed:
Merge pull request #2 from asamuzaK/dev-test-increase

fix increment() behavior and code clean-up

asamuzaK commented:
As I checked comparing version parts,
version expression with `-1` were described as below.

> 1.-1
> < 1 == 1. == 1.0 == 1.0.0
> 
> < 1.1pre10
> < 1.1.-1
> < 1.1 == 1.1.0 == 1.1.00

It can be read to be:
`1.1.-1` is greater than any other expression which contains string part (`pre`, `a` etc.).

From that, I assume `increment ('1.2.-1')` should return `1.2.0`.
So I changed behavior in `increment ('1.2.-1')` to do so.

Also, I've cleaned up some codes.

Sorry for the many try builds, since this was my first GitHub Pull Request.
I think I have done all I can for now in Build #29, so please review.

Regards.

asamuzaK committed:
fix/add increment, decrement parse patterns

asamuzaK committed:
fix parse expression

asamuzaK committed:
fix legacy pattern handling

asamuzaK committed:
fix `pre` match case expression

asamuzaK committed:
fix match pattern

asamuzaK committed:
fix regexp

asamuzaK committed:
fix increment, decrement case

asamuzaK committed:
fix increment case

asamuzaK committed:
try build to check edge cases

asamuzaK committed:
try build 2

asamuzaK committed:
try build 3

asamuzaK committed:
try build 3

asamuzaK committed:
try build 5

asamuzaK committed:
try build 6

asamuzaK committed:
fix from trys

asamuzaK committed:
update test

asamuzaK committed:
fix version string from `pre`

asamuzaK committed:
Merge pull request #3 from asamuzaK/increamt_function_follow_up

handle compare versions, handle legacy versions follow up

asamuzaK commented:
Add follow ups for handling negative numbered versions, legacy versions.

Note: cases like `1.0.a-1 1.0.a-2` are not comparable.
Seems to be a spec bug.
Can not specify `1.0.a-1` to be treated as
`string-b === 'a-' && number-c === '1'`,
or `string-b === 'a' && number-c === '-1'`.
These expression are not used in current Mozilla applications,
but at least matches the toolkit version spec.

`compareVersions('1.0.a-1', '1.0.a-2');`
`compareVersions('1.0.a-2', '1.0.a-1');`
Both returns `-1`.

So, those cases will be thrown in this PR.

asamuzaK committed:
Code optimized

asamuzaK commented:
@jsantell Pull Request is ready.
Please review.

Regards

asamuzaK committed:
Fix some codes and tests for legacy version expression

asamuzaK committed:
Optimized
